### ISSUE TITLE
[ML-03] Fine-tune L2CS-Net for gaze estimation and export l2cs_net.onnx

### ISSUE THREAD

**Issue by victorjudysen** [victorjudysen/ai-exam-proctoring-system#85]
## Context

AI-Based Online Exam Proctoring System — ML Training (Google Colab). Depends on [ML-01]. This ONNX model is consumed by Kweka's [AI-04] gaze estimation endpoint.

### `ml-training/03_gaze_training.ipynb`

**Background:** L2CS-Net is a gaze estimation model that predicts gaze direction angles. We fine-tune it to classify gaze into 5 categories: Screen, Left, Right, Up, Down.

**Cell 1 — Download MPIIFaceGaze dataset:**
The MPIIFaceGaze dataset can be downloaded from:
https://www.mpi-inf.mpg.de/departments/computer-vision-and-machine-learning/research/gaze-based-human-computer-interaction/appearance-based-gaze-estimation-in-the-wild/

```python
# Alternative: use GazeCapture subset or synthetic gaze data
# For FYP scope, MPIIFaceGaze (~3GB) is sufficient
# Dataset format: each sample has a face image + gaze angles (yaw, pitch)
```

**Cell 2 — Map gaze angles to 5 classes:**
```python
def angles_to_class(yaw_deg: float, pitch_deg: float) -> int:
    """
    Maps continuous gaze angles to 5 classes:
    0=Screen, 1=Left, 2=Right, 3=Up, 4=Down
    """
    if abs(yaw_deg) < 15 and abs(pitch_deg) < 15:
        return 0  # Screen
    if yaw_deg < -15:
        return 2  # Right (viewer's right = subject's left yaw is negative)
    if yaw_deg > 15:
        return 1  # Left
    if pitch_deg > 15:
        return 3  # Up
    if pitch_deg < -15:
        return 4  # Down
    return 0
```

**Cell 3 — Model architecture (simplified L2CS-Net):**
```python
import torchvision.models as models
import torch.nn as nn

class GazeClassifier(nn.Module):
    def __init__(self, num_classes=5):
        super().__init__()
        backbone = models.resnet50(pretrained=True)
        # Remove final classification layer
        self.features = nn.Sequential(*list(backbone.children())[:-1])
        self.classifier = nn.Linear(2048, num_classes)

    def forward(self, x):
        x = self.features(x)
        x = x.flatten(1)
        return self.classifier(x)
```

**Cell 4 — Training:**
```python
criterion = nn.CrossEntropyLoss()
optimizer = Adam(model.parameters(), lr=1e-4)
# 15 epochs, batch size 32
# Data augmentation: horizontal flip, brightness jitter, random crop
```

**Cell 5 — ONNX Export:**
```python
dummy_input = torch.randn(1, 3, 448, 448).to(device)
torch.onnx.export(
    model,
    dummy_input,
    '/content/drive/MyDrive/fyp-ai/exports/l2cs_net.onnx',
    input_names=['face_image'],
    output_names=['class_logits'],
    opset_version=14,
    dynamic_axes={'face_image': {0: 'batch_size'}, 'class_logits': {0: 'batch_size'}}
)
```

**Cell 6 — Verify:**
```python
sess = ort.InferenceSession('/content/drive/MyDrive/fyp-ai/exports/l2cs_net.onnx')
dummy = np.random.randn(1, 3, 448, 448).astype(np.float32)
logits = sess.run(['class_logits'], {'face_image': dummy})
print(f"Output shape: {logits[0].shape}")   # Expected: (1, 5)
# Predicted class: np.argmax(logits[0]) → 0=Screen, 1=Left, 2=Right, 3=Up, 4=Down
```

## Input/Output Contract (for Kweka's AI-04)

```
Input:  float32 tensor, shape (1, 3, 448, 448)
        RGB face crop, normalised: mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]

Output: float32 tensor, shape (1, 5)
        Raw logits — apply softmax to get class probabilities
        Classes: [0=Screen, 1=Left, 2=Right, 3=Up, 4=Down]
```

## Quality Target

| Metric | Target |
|--------|--------|
| Test accuracy (5-class) | ≥ 75% |
| Gaze MAE (angle regression, if used) | < 5 degrees |
| Output shape | (1, 5) float32 |

## Sharing

Share `l2cs_net.onnx` from Google Drive with Kweka (same process as [ML-02]).

## Branch

Upload notebook: `ml-training/03_gaze_training.ipynb`
